### PR TITLE
Improve checking of higher quantisers

### LINKED ISSUE #10: `EncodeHQ-CBR` crashses when coding a mid-gray frame
Given a mid gray frame as input, `EncodeHQ-CBR` crashes with a floating point
exception during qindex selection.

[Mid-Grey frame (1 frame, 10 bit, 1280x720,
4:2:2, progressive, YCbCr)](https://github.com/bbc/vc2-reference/files/4823896/mid_grey.zip)

Coding fails as illustrated follows:

    $ EncodeHQ-CBR \
        -v \
        -x 1280 -y 720 --f 4:2:2 -l 10 -r 6 \
        -k LeGall -d 3 -a 2 -u 1 -s $((1280*720*2*2/4)) \
        synthetic_grey.16p2 out.vc2
    input file = synthetic_grey.16p2
    output file = out.vc2
    bytes per sample= 2
    luma depth (bits) = 10
    chroma depth (bits) = 10
    height = 720
    width = 1280
    chroma format = 4:2:2
    interlaced = false
    wavelet kernel = LeGall (5,3) ("LeGall")
    wavelet depth = 3
    vertical slice size (in units of 2**(wavelet depth)) = 1
    horizontal slice size (in units of 2**(wavelet depth)) = 2
    compressed bytes = 921600
    output = Stream
    Vertical slices per picture          = 90
    Horizontal slices per picture        = 80
    Slice bytes numerator                = 128
    Slice bytes denominator              = 1
    Quantisation matrix = 4, 2, 2, 0, 4, 4, 2, 5, 5, 3

    Writing Sequence Header

    Reading input frame number 0
    Forward transform
    Determine quantisation indices
    Floating point exception (core dumped)

### LINKED ISSUE #17: `EncodeHD-CBR` crashes with floating point exception for wavelet test pattern
The following [wavelet test pattern (1 frame, 10 bit, 1280x720, 4:2:2, progressive, YCbCr)](https://github.com/bbc/vc2-reference/files/4824061/wavelet_test_pattern.zip) causes `EncodeHD-CBR` to crash during quantisation index selection.

Example invocation:

    $ EncodeHQ-CBR \
        -x 1280 -y 720 -r 6 -f 4:2:2 -l 10 \
        -k LeGall -d 3 -a 2 -u 1 -s 921600 -S 2 \
        input.16p2 out.vc2 -v
    bytes per sample= 2
    luma depth (bits) = 10
    chroma depth (bits) = 10
    height = 720
    width = 1280
    chroma format = 4:2:2
    interlaced = false
    wavelet kernel = LeGall (5,3) ("LeGall")
    wavelet depth = 3
    vertical slice size (in units of 2**(wavelet depth)) = 1
    horizontal slice size (in units of 2**(wavelet depth)) = 2
    compressed bytes = 921600
    output = Stream
    Vertical slices per picture          = 90
    Horizontal slices per picture        = 80
    Slice bytes numerator                = 128
    Slice bytes denominator              = 1
    Quantisation matrix = 4, 2, 2, 0, 4, 4, 2, 5, 5, 3

    Writing Sequence Header

    Reading input frame number 0
    Forward transform
    Determine quantisation indices
    Floating point exception (core dumped)

Note: The test pattern consists of in-range values (i.e. it is a valid picture)
and is designed to produce near-maximum values in intermediate and output
stages of the wavelet transform. For reference, this image is shown below:

Wavelet test pattern

This file does not appear to crash `EncodeLD`.

### PR DESCRIPTION
**Pull request by GalenReich**
Fixes #10 and fixes #17  

Previously, the checking of higher quantisers would increment q and select the value before the residual was worse than for the smallest possible q. This led to an infinite loop for slices with only zeros, as the residual stayed constant (also at zero). When q incremented beyond 119, the lookup when determining the quantization factor would read off the end of the array. Eventually empty memory would be reached and a division by zero occured, resulting in the error.

Now, the checking of higher quantisers continues as long as there is an improvement in the residual against the previously checked q. If the residual stays constant or gets worse, the checking terminates and the previous q value is used. This prevents the infinite loop when only zeros are present.

Additionally, an error is now thrown if the lookup attempts to read past the end of the array.

### CODE REVIEW

**Comment by GalenReich** [comment]
Also f7a71ac has borked line endings - sorry! It looks like it might be a result of mixed line endings in the repo...

File: src/Library/src/Quantisation.cpp [hunk 1]
```diff
@@ -58,6 +58,7 @@ const int quant_factor(int q) {
 //  0x100000000, 0x1306FE0A3, 0x16A09E668, 0x1AE89F996, 0x200000000, 0x260DFC146, 0x2D413CCD0, 0x35D13F32B
   };
   if (q<0) q=0;
+  if (q>120) q=120;
```
Comment (mossblaser) [inline src/Library/src/Quantisation.cpp]: You can probably avoid having to hand code the array length using ``sizeof(lookup) / sizeof(lookup[0])``

File: src/Library/src/Quantisation.cpp [hunk 2]
```diff
@@ -58,7 +58,7 @@ const int quant_factor(int q) {
 //  0x100000000, 0x1306FE0A3, 0x16A09E668, 0x1AE89F996, 0x200000000, 0x260DFC146, 0x2D413CCD0, 0x35D13F32B
   };
   if (q<0) q=0;
-  if (q>120) q=120;
+  if (q>119) q=119;
```
Comment (mossblaser) [inline src/Library/src/Quantisation.cpp]: `>=120` may be more human-friendly?

File: src/EncodeHQ-CBR/EncodeHQ-CBR.cpp [hunk 3]
```diff
@@ -65,54 +65,56 @@ using std::istream;
...
           trialQ++;
           long long trialYSS = yss_for_slice(slices[row][column], trialQ, qMatrix);
-          if (trialYSS > YSS)
-            break;
-          q = trialQ;
-        } while (true);
+          deltaYSS = trialYSS - prevYSS;
+          prevYSS = trialYSS;
+        } while (deltaYSS <0);
+        q = --trialQ;
```
Comment (mossblaser) [inline src/EncodeHQ-CBR/EncodeHQ-CBR.cpp:114]: Did you mean `q = trialQ  - 1` ?

File: src/Library/src/Quantisation.cpp [hunk 4]
```diff
@@ -39,6 +39,10 @@ const Array2D adjust_quant_indices(const Array2D& qIndices, const int qMatrix) {
 
 const int quant_factor(int q) {
   // Lookup table valid q<120. For q>= 120 quant_factor(q) requires more than 32 bits.
+  if (q>119) {
+    throw std::logic_error("quantization index exceeds 119, corresponding quantization factor exceeds 32 bits.");
```
Comment (mossblaser) [inline src/Library/src/Quantisation.cpp:43]: You can safely omit the "corresponding quantization factor exceeds 32 bits." bit as this is merely an arbitrary internal reason for choosing this limit.

**Pull request merged**